### DISCUSSION TITLE
transform bundle so no runtime parsing is needed

### DISCUSSION THREAD

**Discussion by ivanhofer** [ivanhofer/typesafe-i18n#89] [Optimizations]
In its current implementation `typesafe-i18n` needs to parse each translation the first time it gets called. The parsing part is taking about half of the bundle size. The whole library could be trimmed to < 500 bytes if the parsing would take place compile-time.

Instead of this:

```typescript
const en: Translation = {
   'HELLO': 'Hi {name:string}!'
}
export default en
```

the production bundle output could be:

```typescript
const en: Translation = {
   'HELLO': ['Hi ', { k: 'name' }, '!']
}
export default en
```

This would mean slightly larger output for each translation file, but this should compress (`gzip`) better because of its repetitive pattern. If the parsing is not needed run-time, the library should be even faster (mostly nanoseconds). 

What needs to be done:
 - [ ] evaluate the impact regarding bundle size
 - [ ] evaluate the time-savings if message parsing is not needed
 - [ ] implement an AST-transform function that parses and replaces each message
 - [ ] remove parsing part from output bundle

**Comment by ivanhofer** [comment]
I recently saw this [babel plugin](https://github.com/cibernox/babel-plugin-precompile-intl) that does something similar

**Reply by abastardi** [reply]
Note, [babel-plugin-precompile-intl](https://github.com/cibernox/babel-plugin-precompile-intl) is used by [svelte-intl-precompile](https://github.com/cibernox/svelte-intl-precompile). Maybe have a look at that library for some inspiration as well.

**Comment by Suya1671** [comment]
Maybe make it an [unplugin](https://github.com/unjs/unplugin) thing (unplugin are plugins that work with any bundler)

**Reply by ivanhofer** [reply]
really cool, thanks for sharing!

**Reply by Suya1671** [reply]
> really cool, thanks for sharing!

Svelte-intl-precompile uses a vite plugin, so maybe take inspiration from that

**Reply by ambigos1** [reply]
@Suyashtnt Can I use Svelte-intl-precompile with typesafe-i18n?
This performance optimization is very important. I wonder how now one implemented the optimization since 2021 /:

**Reply by Suya1671** [reply]
> @Suyashtnt Can I use Svelte-intl-precompile with typesafe-i18n? This performance optimization is very important. I wonder how now one implemented the optimization since 2021 /:

Inlang does this natively actually. I'd recommend moving over to that instead since @ivanhofer used to work for it before he (sadly) passed away.

**Reply by ambigos1** [reply]
OH no. It is so sad :(
I will research InLang. I have been working on adapting my websites with typesafe-i18n for over a month.
I hope my work wasn't for nothing
